Add Dockerfile

solsson opened the pull request.

edenhill commented:
Great stuff!

Can you add an example to the README how to use it?

ctrochalakis commented on Dockerfile:
```diff
@@ -0,0 +1,18 @@
+FROM debian:stable-20170620
```
Hello,

Is there a reason you are using the date tag and not just `debian:stable`?

solsson replied:
I thought because the release is so [recent](https://www.debian.org/News/2017/20170617) that I wanted to stress the upgrade. If we remove the date tag I guess that `debian:stretch` makes more sense. Years from now they'll move `:stable` to next release, and the build might break.

edenhill commented:
LGTM.

@ctrochalakis mind a final review?

ctrochalakis commented on Dockerfile:
```diff
@@ -0,0 +1,18 @@
+
+COPY . /usr/src/kafkacat
+
+RUN runDeps=''; \
```
You could drop `runDeps` since it's empty, also, you could avoid setting `-ex` and use ` && ` between commands.

solsson replied:
Is this just a matter of conventions, or does it have any practical implications?

solsson replied:
I'll drop the runDeps. This whole block is just copy paste from a convention I tend to use on debian.

ctrochalakis commented on README.md:
```diff
@@ -145,3 +145,28 @@ Pretty-printed JSON metadata listing
+
+# Running in docker
+
+There is no official image build yet but you may run `docker build -t kafkacat .` or use for example `solsson/kafkacat@sha256:3075a6ca2d8431910cf01af13acd2ef9542c8ed48547733eb5c4321364ef0b66`.
```
It is better not to mention a specific docker image since it will become outdated soon enough.

ctrochalakis commented on README.md:
```diff
@@ -145,3 +145,28 @@ Pretty-printed JSON metadata listing
+
+There is no official image build yet but you may run `docker build -t kafkacat .` or use for example `solsson/kafkacat@sha256:3075a6ca2d8431910cf01af13acd2ef9542c8ed48547733eb5c4321364ef0b66`.
+
+Example:
```
Personally I'd find the example easier to follow if it avoided detaching kafkacat instances. It's closer to how you 'd run kafkacat if it was locally installed, no need to call `docker logs`.

ctrochalakis commented on README.md:
```diff
@@ -145,3 +145,28 @@ Pretty-printed JSON metadata listing
+docker run --rm kafkacat --help
+echo 1 > example.log
+docker run --name test-producer -d -v $(pwd)/example.log:/logs/example.log --entrypoint /bin/bash --net=host kafkacat \
+  -c 'tail -f /logs/example.log | kafkacat -b mybroker -t logs -P'
```
You can do that without passing example.log as a volume (assuming detach is dropped):

`tail -f example.log | docker run -i --name test-producer --net=host kafkacat -b mybroker -t logs -P`

solsson replied:
I just thought this was a more interesting use case, but maybe only because I was working with log aggregation in Kubernetes.

solsson commented:
I'll be on vacation for two weeks and resume this work once I'm back.

solsson commented:
@ctrochalakis thanks for the review. I've committed new suggestions.

Also in celebration of https://github.com/edenhill/librdkafka/releases/tag/v0.11.0 I've built:
```
$ docker run solsson/kafkacat@sha256:1266d140c52cb39bf314b6f22b6d7a01c4c9084781bc779fdfade51214a713a8 help
Error: -b <broker,..> missing

Usage: kafkacat <options> [file1 file2 .. | topic1 topic2 ..]]
kafkacat - Apache Kafka producer and consumer tool
https://github.com/edenhill/kafkacat
Copyright (c) 2014-2015, Magnus Edenhill
Version 1.3.1 (JSON) (librdkafka 0.11.0 builtin.features=gzip,snappy,sasl,regex,lz4,sasl_plain,plugins)
```
and then in b04279c reverted the version change.

edenhill commented on Dockerfile:
```diff
@@ -1,13 +1,16 @@
   \
   cd /usr/src/kafkacat; \
+  echo "Source versions:"; \
+  cat ./bootstrap.sh | grep github_download; \
```
Suggest `grep ^github_download bootstrap.sh`

solsson commented:
> I suspect librdkafka is not getting compiled with openssl or libsasl2 support.

Thanks. Fixed. Do we need `libsasl2-2` and `libssh2-1` to stick around as runtime dependency? Now they're apt-get removed. I don't have an ssl'd kafka to test with.

I also fixed the grep as suggested.

What do you think about 2778479 and 275fade? I think they help outsiders like me produce stable builds, but they also mean that there's another version number to maintain in source, and you don't need it if you plan to publish official docker builds triggered whenever librdkafka is released.

edenhill commented:
I think fixating the versions is a good idea and it isn't too bad to update the 3-4 times librdkafka is released per year.

libsasl2-2 is most likely needed during runtime, libssh* is not afaik.

edenhill commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+  echo "Source versions:"; \
+  grep ^github_download ./bootstrap.sh; \
+  \
+  ./bootstrap.sh; \
```
We'll want to remove tmp-bootstrap before running bootstrap to make sure it gets a clean slate

solsson replied:
21a9146

edenhill commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+  rm -rf /usr/src/kafkacat/tmp-bootstrap; \
+  apt-get purge -y --auto-remove $buildDeps; \
+  rm /var/log/dpkg.log /var/log/alternatives.log /var/log/apt/*.log
+
```
Run `kafkacat -V` to verify that the binary still works after package removal.

solsson replied:
It bailed on lack of libcrypto. I've added all libs now. The resulting layer totals 10.2 MB now instead of 6.6 prior to building with libssl+libsasl.

edenhill replied:
Aye, add `kafkacat -V` as a final step which makes sure all required dependencies are in place.

solsson replied:
ac6c085

solsson commented:
I think we're done. Running it from https://hub.docker.com/r/solsson/kafkacat/ for my use cases, and it works.

edenhill commented:
Great stuff, thanks for this @solsson  !
